rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+  void (*M_Init)(M_context*, string);
+  void (*M_Execute)(M_context*);
+};
+
+// The plugin registry for module M
+// The registry here implemented uses C++ class std::map
+// This is OK as it is not exposed in our example here.
+// As far as module implementations are concerned, they only
+// need to see and use M_RegisterPlugin method, and that method takes // care of their addition into this registry.
+std::map<string, MInterface> m_plugin_registry;
+
+// Function to call for each plugin at load time
+// to add themselves to the registry
+void M_RegisterPlugin(
+    string plugin, 
+    void (*M_Init)(M_context*, string),
+    void (*M_Execute)(M_context*)) {
+  // Omitting error handling.
+  m_plugin_registry.insert(plugin,
+                           MInterface(M_Init, M_Execute));
+}
+
+// Implementation of the interface is just a thin layer to
+// get the correct plugin and call it.
+// Here we assume that plugin explicitly gets selected by
+// the init function. Some modules can go with implicit selection,
+// Such as deducing the filesystem from the file path.
+void M_Init(M_context* ctx, string id) {
+  // Just a quick hacky way to select the plugin here
+  ctx->interface = m_plugin_registry[id];
+  ctx->interface.M_Init(ctx, id);
+}
+
+void M_Execute(M_context* ctx) {
+  ctx->interface.M_Execute(ctx);
+}
+```
+
+
+**Please note that the above is a toy example to explain the abstract ideas described above. Exact implementation can vary across different modules.**
+
+Interface has to be pure C at the ABI level. We can have C++ header-only libraries built on top of these C ABI/APIs.
+
+
+#### Plugins
+
+Plugins need to include implementation of the interfaces declared by one module. If the module interface requires Init and Compute methods, it will need to implement these two functions, plus a TF_InitPlugin function which will be called at load time. This function will also need to register the plugin as prescribed by the module.
+
+
+```
+// A simple plugin implementation A, for module M.
+typedef struct M_context;
+
+// Here is the meat of the plugin:
+void A_Init(M_context* ctx, string id) {
+  // id can be thrown away, in this example.
+  // Or can encode different ways plugin can be initialized.
+  ......
+  // initialize the plugin
+  // Modify the context using functions exposed by core
+  TF_SetFooInContext(ctx, foo);
+  TF_InitBarInContext(ctx);
+}
+
+void A_Execute(M_context* ctx) {
+  ..........
+}
+
+void TF_InitPlugin() {
+  M_RegisterPlugin("A", A_Init, A_Execute);
+}
+```
+
+
+When this plugin is loaded by TF at runtime, `TF_InitPlugin` method will be called. This method will register the plugin as prescribed by the module, and exit.
+
+Plugin shared objects will need to follow some standards:
+
+
+
+1.  These cannot export any symbols into global namespace to avoid symbol collisions.
+1.  They need to communicate to TF through any provided C APIs.
+1.  They can link anything they like, anyway they like in addition to TF.
```

Comment:
How would TF core APIs be called from within a plugin? e.g. If an MKL plugin signals that it can handle a usage, but as part of that usage, the MKL plugin needs to call the eigen implementation in TF core, how would that handoff be managed?

Reply:
This is an interesting use case. I can think of two potential solutions:

1 - MKL plugin can use the global kernel registry to fetch the eigen kernel and execute it.
2 - MKL plugin can rebuild all eigen kernels and also distribute all.

We can brainstorm more on this issue, I think it is certainly worth considering.

Reply:
A global kernel registry sounds interesting. @agramesh1, @nhasabni do you see performance implications here?

Reply:
We already have a global kernel registry in TF:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op_kernel.cc#L1048